Include/README.rst

```diff
@@ -0,0 +1,68 @@
+The Python C API
+================
+
+The C API is divided into three sections:
+
+1. ``Include/``
+2. ``Include/cpython/``
+3. ``Include/internal/``
+
+
+Include: Limited API and Stable ABI
+===================================
+
+``Include/``, excluding the ``cpython`` and ``internal`` subdirectories,
+contains the public Limited API (Application Programming Interface) and
+the Stable ABI (Application Binary Interface).  :pep:`384` defined the
+Stable ABI and the Limited API; two related, but distinct concepts.
+The Stable ABI promises binary compatibility across subsequent versions of
+CPython 3.x, given that the source code restricts itself to the Limited
+API.  The Limited API is a subset of the C API.
+
+Guidelines for expanding the Limited API:
+
+- Functions *must not* steal references
+- Functions *must not* return borrowed references
+- Functions returning references *must* return a strong reference
+- Macros should not expose implementation details
+- Please start a public discussion before expanding the API
+- Functions or macros with a ``_Py`` prefix do not belong in ``Include/``.
+
+
+Include/cpython: CPython implementation details
+===============================================
+
+``Include/cpython/`` contains the public API that is excluded from the
+Limited API and the Stable ABI.  When a new function or macro is
+introduced, it has to be explicitly excluded from the Limited API using
+``#ifndef Py_LIMITED_API``.
```

Comment:
Why is that the case? If everything under `include/cpython` is excluded from the limited API, then individual `#ifdef`s are entirely pointless.

Reply:
A `grep -r Py_LIMITED_API Include/{cpython,internal}` reveals that there's a lot of such exclusions. If everything under `Include/cpython` (and `Include/internal`) is excluded, how is that enforced and implemented? Such information would be valuable in this README, IMO.

Reply:
To me it just hints that the first sentence above is wrong :-)

Reply:
Quoting PEP 652:
```
During the compilation of applications, the preprocessor macro Py_LIMITED_API must be defined.
Doing so will hide all definitions that are not part of the ABI.
```

Reply:
This quote relates to the compilation of applications. It doesn't relate to the organization of CPython headers.

Reply:
All right, I've gone through all the header files, and header files under `Include/cpython` are completely excluded by `#ifndef Py_LIMITED_API` wrappers, either explicitly in the actual file, or implicitly by guarding the `#include` in the corresponding `Include` header. I'll adjust the text.

Reply:
But do note that, for some reason, there are also non-limited API declarations in `Include`:
https://github.com/python/cpython/blob/master/Include/unicodeobject.h#L141

So "the public API that is excluded from the Limited API" is not only in `Include/cpython`, it's also in `Include`.
Which makes the directory organization completely confusing and pointless IMHO.

Reply:
Yes, that's true. I'll see if I can address this and at the same time keep things short and concise.

Thanks for reviewing, @pitrou.